### PR TITLE
Added some same-origin-domain checks

### PR DESCRIPTION
**Pull request by asajeffrey**
This PR adds some same-origin-domain checks, replacing some same-origin checks. This will make a difference once we support setting document.domain.

This PR builds on #15438, only the last commit is part of this PR.

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes do not require tests because we don't support setting document.domain yet

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/15478)

### CODE REVIEW

**Comment by asajeffrey** [comment]
cc @jdm @cbrewster

File: components/script/dom/htmliframeelement.rs [hunk 1]
```diff
@@ -511,15 +497,29 @@ impl HTMLIFrameElementMethods for HTMLIFrameElement {
...
+            None
         }
     }
 
     // https://html.spec.whatwg.org/multipage/#dom-iframe-contentdocument
+    // https://html.spec.whatwg.org/multipage/#concept-bcc-content-document
     fn GetContentDocument(&self) -> Option<Root<Document>> {
-        self.get_content_window().map(|window| window.Document())
+        // Step 1.
+        if let Some(pipeline_id) = self.pipeline_id.get() {
```
Comment (Ms2ger) [inline components/script/dom/htmliframeelement.rs]: Early returns would be better here.

**Review by Ms2ger — COMMENTED** [review]
I'm not convinced it's a good idea to land this until we actually have document.domain, and I can tell you to write tests for each and every change here.

**Comment by asajeffrey** [comment]
I updated the test expectations (https://github.com/servo/servo/pull/15478/commits/d6e2792b20607c9f5688c41b050551478b76ee0a). Summary:

* Some tests now `PASS`, hooray!
```
/html/browsers/history/the-location-interface/security_location_0.htm
/html/browsers/the-window-object/security-window/window-security.html
/html/semantics/embedded-content/the-iframe-element/iframe-load-event.html
/navigation-timing/test_unique_performance_objects.html 
/html/semantics/scripting-1/the-template-element/...
/html/syntax/parsing/template/...
```
* Some tests which were `FAIL` are now `TIMEOUT`. The timeouts are caused by `postMessage`, in particular `event.source` being undefined.
* I bugfixed `tests/wpt/mozilla/tests/mozilla/referrer-policy/generic/common.js`, which was stringifying `iframe.contentWindow.location` even for cross-origin windows, which now generates a security error.

@Ms2ger: with these updates to the test expectations, are you okay with this PR going ahead without `document.domain`?

File: components/script/dom/htmliframeelement.rs (hunk 1 above)
Comment (asajeffrey) [inline components/script/dom/htmliframeelement.rs]: Done.

**Comment by asajeffrey** [comment]
Updated #15438.

**Comment by asajeffrey** [comment]
@Ms2ger I added the setter for document.domain over in #15536, which includes a test for same-origin-domain security checks.

**Comment by asajeffrey** [comment]
@Ms2ger ping?

**Comment by asajeffrey** [comment]
Rebased.

**Comment by asajeffrey** [comment]
Fixed a bug which was causing `blob://http://foo.com/` to be considered dissimilar-origin to `http://foo.com/`. Also rebased.

**Comment by asajeffrey** [comment]
#15438 landed, so this can now land.

@Ms2ger: are you still wanting more tests? If so, we probably need to merge this with #15536.

**Comment by asajeffrey** [comment]
Conversation with @jdm on irc: http://logs.glob.uno/?c=mozilla%23servo&s=24+Feb+2017&e=24+Feb+2017#c619500

TL;DR: merge this PR with #15536.
